### PR TITLE
Optimize integer overflow checks

### PR DESCRIPTION
**Pull request by 0xPhaze**
The arithmetic checks for integer addition and subtraction can be optimized.

Idea:
```solidity
function checkedAddInt(int256 a, int256 b) public pure returns (int256 c) {
    unchecked {
        c = a + b;

        bool overflow;

        assembly {
            // If `a >= 0`, then the sum `c = a + b` can't be less than `b`.
            // If `a <  0`, then the sum `c = a + b` can't be greater than `b`.
            // We combine these two conditions into one using `xor`.
            overflow := xor(slt(a, 0), slt(c, b))
        }

        if (overflow) arithmeticError();
    }
}
```

This saves somewhere between 30-31 gas depending on the optimization setting.

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ Language Features:
 
 
 Compiler Features:
+* Code Generator: More efficient arithmetic checks for integer addition and subtraction.
```
Comment (matheusaaguiar) [inline Changelog.md]:
```suggestion
* Code Generator: More efficient overflow checks for integer addition and subtraction.
```

**Review by matheusaaguiar — COMMENTED** [review]
Looks good!
I would just suggest a minor change in the Changelog entry.

**Review by matheusaaguiar — DISMISSED** [review]
The failing test is unrelated.

**Comment by r0qs** [comment]
@0xPhaze please rebase onto the current develop branch (the fix for `t_native_test_ext_zeppelin` was already merged). Also, please squash the commits ;)

**Comment by 0xPhaze** [comment]
Does this need to be squashed again after the merge?

**Comment by 0xPhaze** [comment]
@r0qs Squashed and rebased.

**Comment by r0qs** [comment]
Hey @0xPhaze thanks! But could you amend the commit message to be a more meaningful description of the change instead of  `squash commits`?

File: Changelog.md [hunk 2]
```diff
@@ -5,7 +5,7 @@ Language Features:
 
 Compiler Features:
  * SMTChecker: Properties that are proved safe are now reported explicitly at the end of the analysis. By default, only the number of safe properties is shown. The CLI option ``--model-checker-show-proved-safe`` and the JSON option ``settings.modelChecker.showProvedSafe`` can be enabled to show the full list of safe properties.
-
+ * Code Generator: More efficient overflow checks for integer addition and subtraction.
```
Comment (matheusaaguiar) [inline Changelog.md]:
```suggestion
 * Code Generator: More efficient overflow checks for integer addition and subtraction.


```

**Review by matheusaaguiar — DISMISSED** [review]
LGTM.

**Review by r0qs — DISMISSED** [review]
I'm approving it as well. But I'm wondering if we need to run the benchmarks to check if anything got improved. What do you think @matheusaaguiar?

**Comment by ekpyron** [comment]
The actual real-world benchmark results on this are actually rather "mixed"... the main thing to think about here is whether the new kinds of checks make it harder to simplify the conditions in combination with other expressions (and whether such simplifications actually happen in the first place). In general, this may be a case where actually splitting the optimizer rule list into two parts, i.e. a "normal form simplification" (tweaked towards bringing expressions into a normal form to collapse complex expressions) and a "finalization" (tweaked towards bringing the normal form into a more complex simpler representation without expecting to combine further with other expressions) pass and making the transition done manually for the overflow checks here part of the "finalization" pass.

The main question is whether we'd still want to merge this as is for now before we'd get to any more complex solution.

### `ir-no-optimize`
|          project |  bytecode_size | deployment_gas | method_gas |
|:----------------:|---------------:|---------------:|-----------:|
|           bleeps |                |                |            |
|            brink |                |                |            |
|        chainlink |                |                |            |
|           colony | **`+0.01% ❌`** |                |            |
|        elementfi |                |                |            |
|              ens |                |                |            |
|            euler |                |                |            |
|           gnosis |                |                |            |
|              gp2 |                |                |            |
|  perpetual-pools |                |                |            |
|    pool-together |                |                |            |
|         prb-math |                |                |            |
|          trident |                |                |            |
|          uniswap |                |                |            |
| yield_liquidator |                |                |            |
|         zeppelin |                |                |            |

### `ir-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps | **`+0.02% ❌`** |           `0%` |           `0%` |
|            brink | **`+0.11% ❌`** |                |                |
|        chainlink | **`-0.01% ✅`** | **`-0.03% ✅`** |          `+0%` |
|           colony |          `+0%` |                |                |
|        elementfi |  **`-0.4% ✅`** |                |                |
|              ens |           `0%` | **`+0.01% ❌`** |           `0%` |
|            euler | **`-0.01% ✅`** | **`-0.02% ✅`** | **`+0.02% ❌`** |
|           gnosis | **`+0.04% ❌`** | **`+0.05% ❌`** | **`+0.04% ❌`** |
|              gp2 | **`+0.04% ❌`** | **`+0.02% ❌`** |           `0%` |
|  perpetual-pools | **`+0.01% ❌`** | **`+0.01% ❌`** | **`-0.01% ✅`** |
|    pool-together | **`+0.03% ❌`** | **`+0.02% ❌`** |          `-0%` |
|         prb-math | **`-0.02% ✅`** | **`-0.04% ✅`** |           `0%` |
|          trident | **`+0.02% ❌`** | **`+0.02% ❌`** | **`+0.01% ❌`** |
|          uniswap |  **`-0.1% ✅`** | **`-0.07% ✅`** | **`-0.11% ✅`** |
| yield_liquidator | **`+0.03% ❌`** | **`+0.01% ❌`** |          `-0%` |
|         zeppelin |           `!V` |           `!V` |           `!V` |

### `ir-optimize-evm-only`
|          project |  bytecode_size | deployment_gas | method_gas |
|:----------------:|---------------:|---------------:|-----------:|
|           bleeps |                |                |            |
|            brink |                |                |            |
|        chainlink |                |                |            |
|           colony | **`+0.01% ❌`** |                |            |
|        elementfi |                |                |            |
|              ens |                |                |            |
|            euler |                |                |            |
|           gnosis |                |                |            |
|              gp2 |                |                |            |
|  perpetual-pools |                |                |            |
|    pool-together |                |                |            |
|         prb-math |                |                |            |
|          trident |                |                |            |
|          uniswap |                |                |            |
| yield_liquidator |                |                |            |
|         zeppelin |                |                |            |

### `legacy-no-optimize`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps |                |                |                |
|            brink | **`+0.07% ❌`** |                |                |
|        chainlink | **`+0.01% ❌`** |                |                |
|           colony | **`+0.01% ❌`** |                |                |
|        elementfi | **`-0.03% ✅`** |                |                |
|              ens |           `0%` |                |                |
|            euler | **`-0.01% ✅`** | **`-0.02% ✅`** | **`+0.01% ❌`** |
|           gnosis | **`+0.02% ❌`** | **`+0.03% ❌`** | **`+0.04% ❌`** |
|              gp2 | **`+0.02% ❌`** |                |                |
|  perpetual-pools |          `+0%` |          `+0%` | **`+0.01% ❌`** |
|    pool-together | **`+0.01% ❌`** | **`+0.01% ❌`** |          `+0%` |
|         prb-math | **`-0.02% ✅`** | **`-0.03% ✅`** |           `0%` |
|          trident | **`+0.01% ❌`** | **`+0.01% ❌`** |          `+0%` |
|          uniswap | **`-0.05% ✅`** | **`-0.04% ✅`** | **`-0.05% ✅`** |
| yield_liquidator | **`+0.01% ❌`** |          `+0%` | **`-0.01% ✅`** |
|         zeppelin |           `!V` |           `!V` |           `!V` |

### `legacy-optimize-evm+yul`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps | **`+0.02% ❌`** |           `0%` |           `0%` |
|            brink |  **`+0.1% ❌`** |                |                |
|        chainlink | **`+0.01% ❌`** | **`-0.01% ✅`** |          `+0%` |
|           colony |          `+0%` |                |                |
|        elementfi | **`-0.07% ✅`** |                |                |
|              ens |           `0%` | **`+0.01% ❌`** |           `0%` |
|            euler | **`-0.05% ✅`** | **`-0.06% ✅`** | **`+0.02% ❌`** |
|           gnosis | **`+0.04% ❌`** | **`+0.04% ❌`** | **`-0.21% ✅`** |
|              gp2 | **`+0.03% ❌`** | **`+0.02% ❌`** |          `-0%` |
|  perpetual-pools |          `+0%` |          `+0%` |          `-0%` |
|    pool-together | **`+0.02% ❌`** | **`+0.02% ❌`** |          `+0%` |
|         prb-math | **`-0.06% ✅`** | **`-0.08% ✅`** |           `0%` |
|          trident | **`+0.02% ❌`** | **`+0.01% ❌`** | **`+0.01% ❌`** |
|          uniswap | **`-0.12% ✅`** |  **`-0.1% ✅`** |  **`-0.1% ✅`** |
| yield_liquidator | **`+0.03% ❌`** |          `+0%` |          `-0%` |
|         zeppelin |           `!V` |           `!V` |           `!V` |

### `legacy-optimize-evm-only`
|          project |  bytecode_size | deployment_gas |     method_gas |
|:----------------:|---------------:|---------------:|---------------:|
|           bleeps |                |                |                |
|            brink | **`+0.09% ❌`** |                |                |
|        chainlink | **`+0.01% ❌`** |          `-0%` |          `+0%` |
|           colony | **`+0.01% ❌`** |                |                |
|        elementfi | **`-0.05% ✅`** |                |                |
|              ens |           `0%` | **`+0.01% ❌`** |           `0%` |
|            euler | **`-0.02% ✅`** | **`-0.03% ✅`** | **`+0.02% ❌`** |
|           gnosis | **`+0.04% ❌`** | **`+0.04% ❌`** | **`-0.21% ✅`** |
|              gp2 | **`+0.03% ❌`** | **`+0.02% ❌`** |          `+0%` |
|  perpetual-pools |          `+0%` |          `+0%` | **`-0.01% ✅`** |
|    pool-together | **`+0.02% ❌`** | **`+0.01% ❌`** |          `-0%` |
|         prb-math | **`-0.04% ✅`** | **`-0.05% ✅`** |           `0%` |
|          trident | **`+0.02% ❌`** | **`+0.01% ❌`** |          `+0%` |
|          uniswap | **`-0.08% ✅`** | **`-0.07% ✅`** | **`-0.09% ✅`** |
| yield_liquidator | **`+0.02% ❌`** |          `-0%` |          `-0%` |
|         zeppelin |           `!V` |           `!V` |           `!V` |


`!V` = version mismatch
`!B` = no value in the "before" version
`!A` = no value in the "after" version
`!T` = one or both values were not numeric and could not be compared
`-0` = very small negative value rounded to zero
`+0` = very small positive value rounded to zero

**Comment by 0xPhaze** [comment]
> The actual real-world benchmark results on this are actually rather "mixed"... the main thing to think about here is whether the new kinds of checks make it harder to simplify the conditions in combination with other expressions (and whether such simplifications actually happen in the first place). In general, this may be a case where actually splitting the optimizer rule list into two parts, i.e. a "normal form simplification" (tweaked towards bringing expressions into a normal form to collapse complex expressions) and a "finalization" (tweaked towards bringing the normal form into a more complex simpler representation without expecting to combine further with other expressions) pass and making the transition done manually for the overflow checks here part of the "finalization" pass.
> 
> The main question is whether we'd still want to merge this as is for now before we'd get to any more complex solution.

Hmm.. I'm actually rather surprised by the results. In all my (albeit simpler) tests I saw it as being cheaper. I can only think of that in the reverting case, the first case would require one additional `xor`. Is there any way for me to further inspect these results? How are these executed/collected?

**Comment by ekpyron** [comment]
We have CI runs compiling several larger real-world projects - the table I posted is generated from the CI artifacts of the ``c_ext_benchmarks`` run (https://app.circleci.com/pipelines/github/ethereum/solidity/28743/workflows/aee21907-f199-4966-aff6-209428d79b0b/jobs/1275995/artifacts) using the script ``scripts/externalTests/benchmark_diff.py`` in the repo. CI basically collects gas stats of the test runs of these projects - the (rather complex) setup for all this is in ``scripts/externalTests``.

But it may very well be that those benchmarks are not actually a particularly good testing surface for the changes in this PR - the external projects tend to be highly optimized and may even largely avoid checked arithmetic altogether. Also these test runs have a non-deterministic variance of ~0.05%, so the results I posted don't necessarily mean things actually really get worse - but the results also don't indicate significant gains.

**Comment by 0xPhaze** [comment]
@ekpyron Only getting back to this now. I only see `c_ext_benchmarks` downloading/comparing benchmarks, but not actually running them. I couldn't find the test fixtures that are being used for the external tests. Are these visible somewhere? I'm also unable to see some files, like the CircleCI config.

In particular, I see the repositories in solidity-external-tests often using older solc versions (0.7.6 for example). How are these tests run and how is the gas/deployment data collected?

**Comment by 0xPhaze** [comment]
Small follow up:

These contracts for example should be entirely unaffected because they specify a pragma version <0.8.0:
- GP2 uses [pragma solidity ^0.7.6](https://github.com/solidity-external-tests/gp2-contracts/blob/main/src/contracts/GPv2Settlement.sol#L2)
- UniswapV2 uses [pragma solidity =0.5.16](https://github.com/solidity-external-tests/uniswap-v2-core/blob/master/contracts/UniswapV2Factory.sol#L1) (perhaps UniswapV3 is meant)

These repositories don't have additional branches with other versions. Perhaps I'm missing something and the script that runs these tests is actually manually replacing the version pragma (which I don't believe is the case). If not, I would expect the code size/costs to remain fixed in the above cases.

Also, in the report there are some cases where code size and deployment/runtime-gas increases **without optimizations**. The changes should reduce both, code size and execution costs if no further optimizations/modifications are applied. Even with optimizations I'm having a hard time thinking of how the optimizer would reduce code size or optimize one branch away.

**Comment by cameel** [comment]
> Perhaps I'm missing something and the script that runs these tests is actually manually replacing the version pragma (which I don't believe is the case).

Yes, our external test framework strips all pragmas and the projects that use pre-0.8 solc are either patched by their runner script ([this is exactly what happens for GP2](https://github.com/ethereum/solidity/blob/v0.8.19/test/externalTests/gp2.sh#L89-L96)) or we have a fork with compatibility patches applied ([this is the case for Uniswap](https://github.com/solidity-external-tests/uniswap-v3-core)), depending how extensive the patches are.

> The changes should reduce both, code size and execution costs if no further optimizations/modifications are applied.

If the difference is small, this could be the result of some non-determinism in tests. For example we often observe very small differences in legacy runs even in PRs that only touch the IR code.

**Comment by ekpyron** [comment]
Getting back to this at least briefly: the main alternative to just merging this PR would be to add the transformation between the verbose checks and the more efficient version as simplification rule. The issue with such a rule, though, is that we potentially make subsequent simplifications that may combine the conditions with further conditions harder - so ideally, we'd *only* run such a simplification rule as a "finalization" step, after we already tried combining expressions with their surrounding. We don't have the setup for that yet, but it has come up multiple times.

So ideally, we'd investigate those options and the effects of changing this in code generation or as simplification rule precisely. Especially, looking at implementing this as a simplification rule in ``RuleList.h`` and checking if that has any *negative* effects (due to making collapsing complex statements further harder, lacking more ``xor``-based-rules) would be nice.

**Comment by 0xPhaze** [comment]
I was not able to replicate most of these results. For example, locally
- Running `./test/externalTests/brink.sh` showed **no** differences in bytecode size
- I was unable to run the tests for Gnosis, as I was getting stack-too-deep (before the PR and after)
- Not sure what happened to prb-math, but all methods show 0 gas (but this is also shown in the comparison report)

Inspecting differences in `method_gas` for Euler showed:
- Comparing this PR to before I got consistently lower gas for `ir-optimize-evm+yul` & `ir-optimize-evm+yul` (didn't bother checking any more configurations), except for one particular function: `EulDistributor::claim(address,address,uint256,bytes32[],address)`
- Re-running the results again, showed that all functions had lower gas
- Comparing two reports, both from before this PR, showed non-determinism and some variance for a few functions, especially the `claim` functions

I was basically hoping to find one case to further inspect, where this would either result in larger bytecode or higher call gas, but couldn't find anything consistent.

TLDR: not sure how much these results can be trusted if there is no fixed seeding or an average across multiple runs.

**Comment by cameel** [comment]
> I was unable to run the tests for Gnosis, as I was getting stack-too-deep (before the PR and after)

Gnosis currently does not build with optimization on the latest compiler. On `develop` we have temporarily disabled IR runs for that test.

> Not sure what happened to prb-math, but all methods show 0 gas (but this is also shown in the comparison report)

A limitation of gas reporter plugin for Hardhat. It does not report gas usage from view methods (because the node does not create an actual transaction when executing them).

> TLDR: not sure how much these results can be trusted if there is no fixed seeding or an average across multiple runs.

That's a good question, that I have no answer for. We never looked into what specifically causes these small non-deterministic changes. Is it the test suite relying on random numbers? Is it fetching mainnet snapshots (that may differ between runs)? Something else? We never had time to properly check and the differences were minimal.

**Comment by ekpyron** [comment]
Since the real-world gas benchmarks on this came out inconclusive and there is reason to prefer a more general rewriting approach to optimize cases like these in general, I'm closing this PR for the time being. We can revisit this issue at a time when improving the optimizer is again our focus.
